miekg/dns
# Server - Graceful shutdown (ensure all queries are processed before closing)

## Summary

Enhance the Shutdown of the server to ensure that all queries received before shutdown is called are completely handled before closing the connections.

related PR : https://github.com/coredns/coredns/pull/1963
related issue : https://github.com/coredns/coredns/issues/1666

### tracking the incoming msg
It requires a SyncGroup to take into account the thread going into READ wait mode, and ack the handle is finished after either an error or the full service of the query.

After shutdown is started none of the listening thread can go start a new READ process. However the current one has to be continued until either a msg is received, either a timeout is reached.

At shutdown, the process wait that all listening thread are terminated before closing the connection

### Timeout
For those who does not want to wait, it is now possible to specify a ShudownTimeout.
Default value is 0 - which means use the biggest of the ReadTimeout of TCP or UDP (incl idleTimeout).


## Tests

I added a test "testLostQueriesAtShutdownServer" that:
- run a server that handle any incoming queries within a random delay of 100 to 600ms
- run a set of 50 concurrent thread. Each one is sending a query to the server until STOP is asked
- wait 1 sec all thread are running
- ask for STOP the msg, 
- wait 50ms that any started sending msg are received bu Server
- ask for Shutdown of the server
- count all sent msg and all received msg by the 50 threads.
- ensure that sent msg == received msg

This test is run for a each of the 3 protocols : UDP, TCP, TLS
